maven-core/src/main/java/org/apache/maven/artifact/transform/ConsumerPomTransformer.java

```diff
@@ -0,0 +1,140 @@
+
+    private final Provider<MavenSession> mavenSessionProvider;
+
+    @Inject
+    public ConsumerPomTransformer( Provider<MavenSession> mavenSessionProvider )
+    {
+        this.mavenSessionProvider = mavenSessionProvider;
+    }
+
+    @Override
+    public DeployRequest transformDeployRequest( RepositorySystemSession session,
+                                                 DeployRequest deployRequest )
+    {
+        MavenSession mavenSession = mavenSessionProvider.get();
+        if ( !Features.buildConsumer( mavenSession.getUserProperties() ).isActive() )
+        {
+            return deployRequest;
+        }
+        return deployRequest.setArtifacts( transformArtifacts( mavenSession, deployRequest.getArtifacts() ) );
+    }
+
+    @Override
+    public InstallRequest transformInstallRequest( RepositorySystemSession session,
+                                                   InstallRequest installRequest )
+    {
+        MavenSession mavenSession = mavenSessionProvider.get();
+        if ( !Features.buildConsumer( mavenSession.getUserProperties() ).isActive() )
+        {
+            return installRequest;
+        }
+        return installRequest.setArtifacts( transformArtifacts( mavenSession, installRequest.getArtifacts() ) );
+    }
+
+    private Collection<Artifact> transformArtifacts( MavenSession mavenSession, Collection<Artifact> artifacts )
+    {
+        TransformerContext context = (TransformerContext) mavenSession
+                .getRepositorySession().getData().get( TransformerContext.KEY );
+        if ( context == null )
+        {
+            return artifacts;
+        }
+        ArrayList<Artifact> result = new ArrayList<>( artifacts.size() );
+        for ( Artifact artifact : artifacts )
+        {
+            if ( "pom".equals( artifact.getExtension() ) )
+            {
+                Path buildOutputDirectory = Paths.get(
+                        mavenSession.getCurrentProject().getBuild().getDirectory() );
+                Path originalPom = artifact.getFile().toPath();
+                Path transformedPom = buildOutputDirectory.resolve( NAME + ".xml" );
+                try
+                {
+                    FileTime originalPomTs = Files.getLastModifiedTime( originalPom );
+                    FileTime transformedPomTs = Files.isRegularFile( transformedPom )
+                            ? Files.getLastModifiedTime( transformedPom ) : null;
+
+                    if ( !originalPomTs.equals( transformedPomTs ) )
+                    {
+                        // save it: either does not exist or TS differ
+                        Files.createDirectories( transformedPom.getParent() );
+                        try ( InputStream inputStream = new ConsumerModelSourceTransformer()
+                                .transform( originalPom, context ) )
+                        {
+                            Files.copy( inputStream, transformedPom );
```

Comment:
One of my personal goals was preventing temporary files, as it opens a backdoor to rewrite the pom.
And I don't think there should be a need for it: passing the inputstream should be possible.

Reply:
Am unsure what backdoor would be solved by not using temporary file. True, here am using component name, but the file name could be randomized instead, and effectively it would prevent any code to tamper with it.

Resolver is all about `java.io.Files`, so no, resolver does not allow passing over InputStream. Moreover, misuse of it's internal APIs (like it happened with FileTransformer), where `org.eclipse.aether.spi.connector.transport.PutTask#dataBytes` (meant and used for checksums only, by resolver itself) was used to transfer rewritten POM is just IMHO merely wrong: it misuses resolver API to prevent Maven API misuse....?

This whole "do not save the file" to me looks like some sort of "premature optimization" parallel: it is assumed, that not saving transformed POM prevents some sort of backdoor, that we are actually not even sure about what it is....  Let's solve problems when we actually hit them IMHO.

Finally, IMHO all this feature is really Maven concern, not Resolver concern. And true, maven-3.x does not have any "callback" to transform "just in time" when m-install-p or m-deploy-p calls (directly) into resolver, but IMHO this is much cleaner solution, and you are using proper APIs just to "swap out" (replace) the artifact about to be deploy to another one. Also, it works, as ITs are showing. Hence, I think this should be solved within realm of Maven, and not have Resolver involved into this at all.

Reply:
Overall I agree and main thing is that the concept of build/consumer keeps working. 
However, I think it is worth reconsidering if resolver is really about Files and not InputStreams. With the architecture of Maven 2+3 it makes sense to use File as the base, but IMO with Maven 4+5 an InputStream would make more sense. And as a bonus I would expect less IO which should result in better performance as well.

Reply:
Not to forget upcoming Maven API, that also (may) solve this issue. Ultimate goal is to confine Resolver fully into Maven, and unlike today (just think about sonatype/eclipse package change) not have it littered into plugins, but offer API for all the use cases plugins would do, and make them "come back to core" instead, where all kind of things would become possible, and would be safe as well from future changes like possibly eclipse/apache package change of resolver for example... (not that it's about to happen tomorrow, but once in future will for sure).